arangod/Aql/ClusterNodes.cpp

```diff
@@ -74,6 +77,662 @@ arangodb::velocypack::StringRef toString(GatherNode::SortMode mode) noexcept {
+
+    while (skipped == 0 && _atDep < _dependencies.size() - 1) {
+      _atDep++;
+      skipped = _dependencies[_atDep]->skipSome(atMost);
+    }
+
+    _done = (skipped == 0);
+
+    return skipped;
+
+    // cppcheck-suppress style
+    DEBUG_END_BLOCK();
+  }
+
+ private:
+  /// @brief _atDep: currently pulling blocks from _dependencies.at(_atDep),
+  size_t _atDep{};
+}; // UnsortingGatherBlock
+
+////////////////////////////////////////////////////////////////////////////////
+/// @struct SortingStrategy
+////////////////////////////////////////////////////////////////////////////////
+struct SortingStrategy {
+  typedef std::pair<
+    size_t, // dependecy index
+    size_t // position within a dependecy
+  > ValueType;
+
+  virtual ~SortingStrategy() = default;
+
+  /// @brief returns next value
+  virtual ValueType nextValue() = 0;
+
+  /// @brief prepare strategy fetching values
+  virtual void prepare(std::vector<ValueType>& /*blockPos*/) { }
+
+  /// @brief resets strategy state
+  virtual void reset() = 0;
+}; // SortingStrategy
+
+////////////////////////////////////////////////////////////////////////////////
+/// @class HeapSorting
+/// @brief "Heap" sorting strategy
+////////////////////////////////////////////////////////////////////////////////
+class HeapSorting final : public SortingStrategy, private OurLessThan  {
+ public:
+  HeapSorting(
+      arangodb::transaction::Methods* trx,
+      std::vector<std::deque<AqlItemBlock*>>& gatherBlockBuffer,
+      std::vector<SortRegister>& sortRegisters) noexcept
+    : OurLessThan(trx, gatherBlockBuffer, sortRegisters) {
+  }
+
+  virtual ValueType nextValue() override {
+    TRI_ASSERT(!_heap.empty());
+    std::push_heap(_heap.begin(), _heap.end(), *this); // re-insert element
+    std::pop_heap(_heap.begin(), _heap.end(), *this); // remove element from _heap but not from vector
+    return _heap.back();
+  }
+
+  virtual void prepare(std::vector<ValueType>& blockPos) override {
+    TRI_ASSERT(!blockPos.empty());
+
+    if (_heap.size() == blockPos.size()) {
+      return;
+    }
+
+    _heap.clear();
+    std::copy(blockPos.begin(), blockPos.end(), std::back_inserter(_heap));
+    std::make_heap(_heap.begin(), _heap.end()-1, *this); // remain last element out of heap to maintain invariant
+    TRI_ASSERT(!_heap.empty());
+  }
+
+  virtual void reset() noexcept override {
+    _heap.clear();
+  }
+
+  bool operator()(
+      std::pair<size_t, size_t> const& lhs,
+      std::pair<size_t, size_t> const& rhs
+  ) const {
+    return OurLessThan::operator()(rhs, lhs);
+  }
+
+ private:
+  std::vector<std::reference_wrapper<ValueType>> _heap;
+}; // HeapSorting
+
+////////////////////////////////////////////////////////////////////////////////
+/// @class MinElementSorting
+/// @brief "MinElement" sorting strategy
+////////////////////////////////////////////////////////////////////////////////
+class MinElementSorting final : public SortingStrategy, public OurLessThan {
```

Comment:
Why does this derive publicly from OurLessThan, whereas HeapSorting derives privately from OurLessThan?
Is it intentional?

Reply:
yes, I wanted to express the fact that HeapSorting hides derived `operator()`

Reply:
vale